### ISSUE TITLE
[BUG] 社区有同学测试发现GLM2每次非训练运行时logist值会变，这里也许有一个BUG待官方同学确认，

### ISSUE THREAD

**Issue by lilongxian** [THUDM/ChatGLM2-6B#305]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

 现象问题来自于：issue 302 

### Steps To Reproduce

这种每次推理得到的logist都会变，很大可能就在于dropout、L2等的使用上设置了固定值，没有区分训练和非训练的状态。
，我在源码中找到了一处，如下：
class ChatGLMModel(ChatGLMPreTrainedModel):
。。。。。
if self.pre_seq_len is not None:
............
self.dropout = torch.nn.Dropout(0.1)

def get_prompt(self, batch_size, device, dtype=torch.half):

   .........................
 
    past_key_values = self.dropout(past_key_values)

    return past_key_values

以上代码中的dropout被用在了前置编码器词嵌入方面：
past_key_values = self.embedding(prefix)
这是一个很重要的词嵌入，但是dropout比率被写死，训练中倒是应该维0.1，但在验证和预测推理应用中，无疑这个dropout=0.1会没运行一次模型裁剪你的前置编码词嵌入矩阵一次参数，你的logist必定会变！！！

目前解决办法：

等待官方确认。
自己下载源码，自己修改一下，仅当训练时维0.1，非训练时为0.0即可，尝试。
最简单的：放弃使用GLM2前置编码器即可，办法：将 --pre_seq_len参数改为None或者0
希望对你有帮助，欢迎大家踊跃讨论，一起为清华智谱GLM开源项目加油！

### Environment

```markdown
- OS:
- Python:
- Transformers:
- PyTorch:
- CUDA Support (`python -c "import torch; print(torch.cuda.is_available())"`) :
```

**Comment by binary-husky** [comment]
pre_seq_len是给下游任务ptuning v2指令微调用的，需要设为0

**Comment by mzh1996** [comment]
> pre_seq_len是给下游任务ptuning v2指令微调用的，需要设为0

您好，我是问题的发现者。我是在eval模式下读取的模型，可能不涉及您说的这个原因。另外，我把模型路径换位ChatGLM1的，就没有这个问题。具体见
[https://github.com/THUDM/ChatGLM2-6B/issues/302](url)

**Comment by lilongxian** [comment]
这就对了，pre_seq_len是给下游任务ptuning v2指令微调用的，需要设为0 ，这就等价于禁用前置位置编码器了。这是最快的办法。

但是推理中不禁用的话，在前置编码器初始化past_key_values时候会dropout掉一点（KV前置位置词嵌入权重的）参数，我觉得这块是不是改为可选dropout更好呢，，，，

**Comment by duzx16** [comment]
Duplicate of #302

**Comment by Hoonly** [comment]
torch的dropout内部会区分当前是训练还是推理，如果你设置了model.eval()，dropout是不会执行的，具体可以看一下torch的源码

**Comment by libreliu** [comment]
> torch的dropout内部会区分当前是训练还是推理，如果你设置了model.eval()，dropout是不会执行的，具体可以看一下torch的源码

`torch.nn.functional.scaled_dot_product_attention` 据我的测试是有问题的（`torch 2.0.1+cu117`），并不会认 model.eval()，参考 https://github.com/THUDM/ChatGLM2-6B/issues/294#issuecomment-1690914917

**Comment by Hoonly** [comment]
这是来自QQ邮箱的自动回复邮件。
 
您好，您的邮件我已收到，谢谢。
